File: TP/TP01/Q08 e Q09 - Arquivo/src/Iterativo.java
import java.io.RandomAccessFile;
import java.util.Scanner;

public class Iterativo {
    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(); // Número de valores reais a serem lidos
        RandomAccessFile arq = new RandomAccessFile("numeros.txt", "rw");

        // Ler os valores e escrever no arquivo
        for (int i = 0; i < n; i++) {
            double valor = sc.nextDouble();
            arq.writeDouble(valor);
        }
        
        // Fechar o arquivo após a escrita
        arq.close();

        // Reabrir o arquivo para leitura
        arq = new RandomAccessFile("numeros.txt", "r");

        // Ler os valores de trás para frente
        for (int i = n - 1; i >= 0; i--) {
            arq.seek(i * 8); // Cada double ocupa 8 bytes
            double valor = arq.readDouble();

            // Verifica se a parte decimal é zero
            if (valor % 1 == 0) {
                System.out.println((int) valor); // Imprime como inteiro
            } else {
                System.out.println(valor); // Imprime como double
            }
        }

        // Fechar o arquivo após a leitura
        arq.close();
        sc.close();
    }
}

File: TP/TP01/Q08 e Q09 - Arquivo/iterativo.c
#include <stdio.h>

int main()
{
    int n;
    double valor;
    FILE *arq;

    // Ler o número de valores reais a serem lidos
    scanf("%d", &n);

    // Abrir o arquivo para escrita
    arq = fopen("numeros.txt", "w");
    if (arq == NULL)
    {
        return 1; // Termina o programa caso o arquivo não possa ser aberto
    }

    // Ler os valores e escrever no arquivo
    for (int i = 0; i < n; i++)
    {
        scanf("%lf", &valor);
        fwrite(&valor, sizeof(double), 1, arq);
    }

    // Fechar o arquivo após a escrita
    fclose(arq);

    // Reabrir o arquivo para leitura
    arq = fopen("numeros.txt", "r");
    if (arq == NULL)
    {
        return 1; // Termina o programa caso o arquivo não possa ser aberto
    }

    // Ler os valores de trás para frente
    for (int i = n - 1; i >= 0; i--)
    {
        fseek(arq, i * sizeof(double), SEEK_SET); // Calcula a posição de leitura e posiciona o ponteiro diretamente no início do valor i-ésimo no arquivo.
        fread(&valor, sizeof(double), 1, arq);

        printf("%g\n", valor); // Imprime como double usando o formato mais compacto e adequado ( inteiro se x.0 ou double se real )
    }

    // Fechar o arquivo após a leitura
    fclose(arq);

    return 0;
}
